stan/math/prim/scal/fun/grad_reg_lower_inc_gamma.hpp

```diff
@@ -0,0 +1,189 @@
+ * Gautschi's formulation appears to be sufficiently accurate
+ * everywhere except for large z and small to moderate a. The
+ * intersection between the two regions is a radius 12 quarter circle
+ * centered at a=0, z=30 although both implementations are
+ * satisfactory near the intersection.
+ *
+ * Some limits that could be treated, e.g., infinite z should
+ * return tgamma(a) * digamma(a), throw instead to match the behavior of,
+ * e.g., boost::math::gamma_p
+ *
+ * @tparam T1 type of a
+ * @tparam T2 type of z
+ * @param[in] a shared with complete Gamma
+ * @param[in] z value to integrate up to
+ * @param[in] precision series terminates when increment falls below
+ * this value.
+ * @param[in] max_steps number of terms to sum before throwing
+ * @throw std::domain_error if the series does not converge to
+ * requested precision before max_steps.
+ *
+ */
+template <typename T1, typename T2>
+typename return_type<T1, T2>::type grad_reg_lower_inc_gamma(
+    const T1& a, const T2& z, double precision = 1e-10, int max_steps = 1e5) {
+  using std::exp;
+  using std::log;
+  using std::pow;
+  typedef typename return_type<T1, T2>::type TP;
+
+  if (is_nan(a))
+    return std::numeric_limits<TP>::quiet_NaN();
+  if (is_nan(z))
+    return std::numeric_limits<TP>::quiet_NaN();
+  if (a > 0 && is_inf(a))
+    domain_error("d_lower_reg_inc_gamma_da", "a", a, "The value ",
+                 " is not finite.");
+  if (z > 0 && is_inf(z))
+    domain_error("d_lower_reg_inc_gamma_da", "z", z, "The value ",
+                 " is not finite.");
+  if (a < 0)
+    domain_error("d_lower_reg_inc_gamma_da", "a", a, "The value ",
+                 " can not be negative.");
+  if (z < 0)
+    domain_error("d_lower_reg_inc_gamma_da", "z", z, "The value ",
+                 " can not be negative.");
+
+  if (z == 0.0)
+    return 0.0;
+
+  if ((a < 0.8 && z > 15.0) || (a < 12.0 && z > 30.0)
+      || a < sqrt(-756 - value_of_rec(z) * value_of_rec(z)
+                  + 60 * value_of_rec(z))) {
+    T1 tg = tgamma(a);
+    T1 dig = digamma(a);
+    return -grad_reg_inc_gamma(a, z, tg, dig, max_steps, precision);
+  }
+
+  T2 log_z = log(z);
+  T2 emz = exp(-z);
+
+  int n = 0;
+  TP sum_a = 0.0;
+
+  T1 a_plus_n = a;  // n begins at zero.
+  T1 lgamma_a_plus_1 = lgamma(a + 1);
+  T1 lgamma_a_plus_n_plus_1 = lgamma_a_plus_1;
+  TP term;
+  while (true) {
+    term = exp(a_plus_n * log_z - lgamma_a_plus_n_plus_1);
+    sum_a += term;
+    if (term <= precision)
+      break;
+    if (n >= max_steps)
+      domain_error("d_lower_reg_inc_gamma_da", "n (internal counter)",
+                   max_steps, "exceeded ",
+                   " iterations, gamma_p(a,z) gradient (a) "
+                   "did not converge.");
+    ++n;
+    lgamma_a_plus_n_plus_1 += log1p(a_plus_n);
+    ++a_plus_n;
+  }
+
+  n = 1;
+  TP sum_b = digamma(a + 1) * exp(a * log_z - lgamma_a_plus_1);
+  a_plus_n = a + 1;  // n begins at one.
```

Comment:
move this one line up and remove the comment. Or... just add `n`? It's an `int`, so there's no speed benefit here.